### PR TITLE
TT-10387 Connector handler

### PR DESCRIPTION
**Pull request by tbuchaillot**
## Description

This PR Implements Connector interface for temporal storage. The main idea come from making connections reusable across different storages. 

For example, in order to reuse a connection between a KeyValue and a List it would be something like:
```go

config := &types.RedisOptions{Addrs: []string{"localhost:6379"}}
conn, err := NewConnector(WithRedisConfig(config))
if err != nil {
    ...
}

kv, _ := NewKeyValue(conn)
list, _ := NewList(conn)

```

The connector interface is:
```go
type Connector interface {
	// Disconnect disconnects from the backend
	Disconnect(context.Context) error

	// Ping executes a ping to the backend
	Ping(context.Context) error

	// Type returns the  connector type
	Type() string

	// As converts i to driver-specific types.
	As(interface{}) bool
}
```

And special mention here to the method `As` - the idea is 100% inspired on https://gocloud.dev/concepts/as/.
Each driver that receives a connection handler , will need to use this in its New func in order to obtain the internal client type.
For example, in case of a KeyValue driver it would something like:
```go
func NewKeyValue( conn Connector) KV {
        var client redis.UniversalClient
	if ok := conn.As(&client); !ok {
              // error here - you are trying to use a different connector in this driver
        }
       
         client.Ping(context.Background()).Result()
}

```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)
- [ ] Documentation updates or improvements.

## Checklist

- [ ] I have [reviewed the guidelines](./CONTRIBUTING.md) for contributing to this repository.
- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [ ] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
     - [ ] I have manually updated the README(s)/documentation accordingly.
     - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
    - [ ] `gofmt -s -w .`
    - [ ] `go vet ./...`

### CODE REVIEW

File: temporal/connector/handler_test.go
Comment (mativm02) [inline temporal/connector/handler_test.go:1]: change file's name to connector_test

File: .DS_Store
Comment (mativm02) [inline .DS_Store:1]: remove this file

**Review by mativm02 — APPROVED** [review]
amazing job 🚀

File: temporal/connector/internal/driver/redisv8/redisv8.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	return h.client.Ping(ctx).Err()
+}
+
+func (h *Connector) Type() string {
+	return types.RedisV8Type
+}
+
+// As converts i to driver-specific types.
+// redisv8 connector supports only *redis.UniversalClient.
+func (h *Connector) As(con interface{}) bool {
```
Comment (sredxny) [inline temporal/connector/internal/driver/redisv8/redisv8.go:93]: can we have a more descriptive name for this function? The name doesn;t describe what it does

**Review by sredxny — APPROVED** [review]
approved with some comments

File: temporal/connector/handler_test.go
Comment (tbuchaillot) [reply temporal/connector/handler_test.go:1]: renamed

File: .DS_Store
Comment (tbuchaillot) [reply .DS_Store:1]: it's removed :D

File: temporal/connector/internal/driver/redisv8/redisv8.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		client = redis.NewClusterClient(universalOpts.Cluster())
+	default:
+		client = redis.NewClient(universalOpts.Simple())
+	}
+
+	return &Connector{client: client, cfg: opts}, nil
+}
+
+func (h *Connector) IsConnected() {
+}
```
[the remark this answers is not in the document]
Comment (tbuchaillot) [reply temporal/connector/internal/driver/redisv8/redisv8.go]: removed the whole function @sredxny  - I'll add it back when we work on automatic reconnection.

File: temporal/connector/internal/driver/redisv8/redisv8.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+func (h *Connector) Ping(ctx context.Context) error {
+	return h.client.Ping(ctx).Err()
+}
+
+func (h *Connector) Type() string {
+	return types.RedisV8Type
+}
+
+// As converts i to driver-specific types.
```
[the remark this answers is not in the document]
Comment (tbuchaillot) [reply temporal/connector/internal/driver/redisv8/redisv8.go:88]: renamed the param - `i` is for `interface{}`

File: temporal/connector/internal/driver/redisv8/redisv8.go (hunk 1 above)
Comment (tbuchaillot) [reply temporal/connector/internal/driver/redisv8/redisv8.go]: Improved the description

**Pull request merged**